compiler/src/iree/compiler/Codegen/Transforms/Transforms.cpp

```diff
@@ -473,5 +475,122 @@ void populateRemoveDeadMemAllocPatterns(RewritePatternSet &patterns) {
+    for (size_t i : llvm::seq<size_t>(0, groups.size())) {
+      AllocGroup &group = groups[i];
+      for (Operation *user : alloc->getUsers()) {
+        // Skip the whole analysis if any user is a subview.
+        // TODO: This could be extended if needed by recursively merging
+        // liveness.
+        if (isa<memref::SubViewOp>(user)) return;
+        if (group.liveness.count(user)) {
+          aliasGroups.push_back(i);
+          break;
+        }
+      }
+    }
+    if (aliasGroups.empty()) {
+      // If we didn't find any alias group create a new one.
+      AllocGroup &newGroup = groups.emplace_back();
+      newGroup.allocs.push_back(alloc);
+      Liveness::OperationListT liveInfo =
+          liveness.resolveLiveness(alloc->getResult(0));
+      newGroup.liveness.insert(liveInfo.begin(), liveInfo.end());
+    } else {
+      // Merge the alloc into the first alias group it interfers with.
+      AllocGroup &mergeGroup = groups[aliasGroups[0]];
+      mergeGroup.allocs.push_back(alloc);
+      Liveness::OperationListT liveInfo =
+          liveness.resolveLiveness(alloc->getResult(0));
+      mergeGroup.liveness.insert(liveInfo.begin(), liveInfo.end());
+      // Then merge all the other alias groups into the first group.
+      for (size_t i = 1, e = aliasGroups.size(); i < e; i++) {
+        AllocGroup &group = groups[aliasGroups[i]];
+        mergeGroup.allocs.insert(mergeGroup.allocs.end(), group.allocs.begin(),
+                                 group.allocs.end());
+        mergeGroup.liveness.insert(group.liveness.begin(),
+                                   group.liveness.end());
+        // For simplicity we leave the group empty and don't remove it.
+        group.allocs.clear();
+        group.liveness.clear();
+      }
+    }
+  }
+
+  LLVM_DEBUG({
+    for (size_t i = 0; i < groups.size(); i++) {
+      llvm::dbgs() << "Alias group " << i << ":\n";
+      for (Operation *op : groups[i].allocs) op->dump();
+    }
+  });
+
+  for (size_t i = 0; i < groups.size(); i++) {
+    if (groups[i].allocs.empty()) continue;
+    aliasGroups.push_back(std::move(groups[i].allocs));
+  }
+}
+
+static int64_t getAllocSize(Operation *op, DataLayout &dataLayout) {
+  auto allocOp = cast<memref::AllocOp>(op);
+  int64_t numElements = allocOp.getType().getNumElements();
+  return (dataLayout.getTypeSizeInBits(allocOp.getType().getElementType()) *
+          numElements) /
+         8;
+}
+
+void packAllocs(OpBuilder &builder, func::FuncOp funcOp,
+                ArrayRef<AliasGroup> aliasGroups) {
+  DataLayout dataLayout = DataLayout::closest(funcOp);
+  builder.setInsertionPointToStart(&(*funcOp.getBody().begin()));
+  int64_t maxAlloc = 0;
+  for (size_t i = 0; i < aliasGroups.size(); i++) {
+    int64_t allocSize = 0;
+    for (Operation *alloc : aliasGroups[i]) {
+      allocSize += getAllocSize(alloc, dataLayout);
+    }
+    maxAlloc = std::max(maxAlloc, allocSize);
+  }
+
+  Attribute memorySpace = aliasGroups[0][0]
```

Comment:
check for the alias group being empty. Also not sure we want to coalesce all alias group into a single allocation? I thought each alias group gets its own allocation.... (sorry still having a hard time navigating what the names of the different data structures are, and the information they carry)

Reply:
If each alias group has its own allocation I don't get to re-use memory/reduce memory usage which is really what I'm trying to do here.